### DISCUSSION TITLE
Out of Order Execution

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#2663] [Language Ideas] [answered]
Hi All,  I just had a thought that I wanted to share.

I know that with myself, I often write code in a logical order and not necessarily the order that I need the data.  For example, I might write the following code:

````
            var Staff = UsersToGenerate.Where(...).ToLinkedList();
            var Attorneys = UsersToGenerate.Where(...).ToLinkedList();

            GenerateReport_Staff(Staff, Context);
            GenerateReport_Attorneys(Staff, Context);
````

But really the following code would be better:
````
            var Staff = UsersToGenerate.Where(...).ToLinkedList();
            GenerateReport_Staff(Staff, Context);

            var Attorneys = UsersToGenerate.Where(...).ToLinkedList();
            GenerateReport_Attorneys(Staff, Context);
````
because it would defer the execution of Attorneys until after Staff is executed (and potentially GCed).

I would like to propose that the C# compiler be allowed to rearrange independent, side-effect free statements so that execution is deferred whenever possible.

**Comment by YairHalberstadt** [comment, marked as answer]
> I would like to propose that the C# compiler be allowed to rearrange independent, side-effect free statements so that execution is deferred whenever possible.

The C# compiler is already allowed to do anything that has no observable effects other than its effect on performance. That would include rearranging independent, side-effect free statements so that execution is deferred whenever possible.

Unfortunately proving that statements are independent and side-effect free is almost impossible in C#, for all but the most trivial of cases.
